src/gurt/dyn_hash.c

```diff
@@ -0,0 +1,1269 @@
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	if (!(htable->ht_ops.hop_key_get(item, &key, &ksize))) {
+		D_ERROR("Get key function failed\n");
+		D_GOTO(out, rc = false);
+	}
+	siphash = gen_siphash(key, ksize);
+	rc = dyn_hash_rec_delete(gtable, (const void *)key, ksize, siphash);
+out:
+	return rc;
+}
+
+bool
+dyn_hash_rec_evict(struct d_hash_table *htable, const void *key,
+		   unsigned int ksize)
+{
+	return false;
+}
+
+bool
+dyn_hash_rec_evict_at(struct d_hash_table *htable, dh_item_t item,
+		      uint64_t siphash)
+{
+	return false;
+}
+
+void
+dyn_hash_rec_addref(struct d_hash_table *gtable, dh_item_t item)
+{
+	struct dyn_hash *htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+	htable->ht_write_lock(htable);
+	htable->ht_ops.hop_rec_addref(gtable, item);
+	htable->ht_rw_unlock(htable);
+}
+
+void
+dyn_hash_rec_decref(struct d_hash_table *gtable, dh_item_t item)
+{
+	struct dyn_hash *htable = gtable->dyn_hash;
+	bool		zombie;
+	bool		ephemeral = (gtable->ht_feats & D_HASH_FT_EPHEMERAL);
+	void		*key;
+	unsigned int	ksize;
+	uint64_t	siphash;
+	uint32_t	index;
+	dh_bucket_t	*bucket;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	htable->ht_write_lock(htable);
+	zombie = htable->ht_ops.hop_rec_decref(gtable, item);
+	if (zombie) {
+		if (!htable->ht_ops.hop_key_get(item, &key, &ksize)) {
+			D_ERROR("Get key function failed\n");
+			htable->ht_rw_unlock(htable);
+			D_GOTO(out, 0);
+		}
+		siphash = gen_siphash(key, ksize);
+		index = (uint32_t)(siphash >> htable->ht_shift);
+		bucket = htable->ht_vector.data[index];
+		htable->bucket_lock(bucket);
+		do_delete(gtable, key, ksize, siphash, bucket, index);
+		if (ephemeral) {
+			/* call user operation */
+			htable->ht_ops.hop_rec_free(gtable, item);
+		}
+	} else {
+		htable->ht_rw_unlock(htable);
+	}
+out:
+	;
+}
+
+int
+dyn_hash_rec_ndecref(struct d_hash_table *gtable, int count, dh_item_t item)
+{
+	bool		zombie = false;
+	int		rc = 0;
+	struct dyn_hash *htable = gtable->dyn_hash;
+	bool		ephemeral = (gtable->ht_feats & D_HASH_FT_EPHEMERAL);
+	void		*key;
+	unsigned int	ksize;
+	uint64_t	siphash;
+	uint32_t	index;
+	dh_bucket_t	*bucket;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	htable->ht_write_lock(htable);
+	if (htable->ht_ops.hop_rec_ndecref != def_hop_ndecref) {
+		rc = htable->ht_ops.hop_rec_ndecref(gtable, item, count);
+		if (rc >= 1) {
+			zombie = true;
+			rc = 0;
+		}
+	} else {
+		do {
+			zombie = htable->ht_ops.hop_rec_decref(gtable,
+				item);
+		} while (--count && !zombie);
+
+		if (count != 0) {
+			/* hash has no records */
+			rc = -DER_INVAL;
+		}
+	}
+	if (rc == 0 && zombie) {
+		if (!htable->ht_ops.hop_key_get(item, &key, &ksize)) {
+			D_ERROR("Get key function failed\n");
```

Comment:
same as above regarding error, need to print more information to be able to debug this if it fails